From 489c2ee9ebdec59a7c4f73d288e1e89873adb2c0 Mon Sep 17 00:00:00 2001
From: Anthony Lopez <anthlope@codeaurora.org>
Date: Wed, 10 Mar 2021 19:38:14 -0800
Subject: [PATCH 1/1] User-specified nanosecond interval patch

Add nanosecond interval to the list of command line arguments. If provided, the
packet interval will use nanosecond precision instead of millisecond.

Signed-off-by: Anthony Lopez <anthlope@codeaurora.org>
---
 src/udpreplay.cpp | 44 +++++++++++++++++++++++++++++++++++---------
 1 file changed, 35 insertions(+), 9 deletions(-)

diff --git a/src/udpreplay.cpp b/src/udpreplay.cpp
index f90aee7..be54ed6 100644
--- a/src/udpreplay.cpp
+++ b/src/udpreplay.cpp
@@ -20,7 +20,9 @@ int main(int argc, char *argv[]) {
   int ifindex = 0;
   int loopback = 0;
   double speed = 1;
-  int interval = -1;
+  int interval = -1; //millisec
+  int nano_interval = -1; // nanosec
+  double rate = 0.0;
   int repeat = 1;
   int ttl = -1;
   int broadcast = 0;
@@ -29,7 +31,7 @@ int main(int argc, char *argv[]) {
   int dest_port = 0;
 
   int opt;
-  while ((opt = getopt(argc, argv, "i:bls:c:r:t:d:g:p:")) != -1) {
+  while ((opt = getopt(argc, argv, "i:bls:c:n:r:t:d:g:p:")) != -1) {
     switch (opt) {
     case 'i':
       ifindex = if_nametoindex(optarg);
@@ -49,10 +51,27 @@ int main(int argc, char *argv[]) {
       break;
     case 'c':
       interval = std::stoi(optarg);
-      if (interval < 0) {
+      if (interval <= 0) {
         std::cerr << "interval must be non-negative integer" << std::endl;
         return 1;
       }
+      rate = (double)interval/1000.0;
+      printf("Packets will be sent with interval of %d ms or %.2f s\n",
+           interval, rate);
+      printf("Rate will be %.2f packets / sec\n",
+            1.0/rate);
+      break;
+    case 'n':
+      nano_interval = std::stoi(optarg);
+      if(nano_interval <= 0) {
+        std::cerr << "nanosec interval must be non-negative integer" << std::endl;
+        return 1;
+      }
+      rate = (double)nano_interval/(double)NANOSECONDS_PER_SECOND;
+      printf("Packets will be sent with interval of %d ns or %.2f s\n",
+           nano_interval, rate);
+      printf("Rate will be %.2f packets / sec\n", 
+            1.0/rate);
       break;
     case 'r':
       repeat = std::stoi(optarg);
@@ -97,13 +116,14 @@ int main(int argc, char *argv[]) {
     std::cerr
         << "udpreplay 1.0.0 © 2020 Erik Rigtorp <erik@rigtorp.se> "
            "https://github.com/rigtorp/udpreplay\n"
-           "usage: udpreplay [-i iface] [-l] [-s speed] [-c millisec] [-r "
-           "repeat] [-t ttl] [-d dest ipv4] [-g dest ipv6] [-p destination port] "
+           "usage: udpreplay [-i iface] [-l] [-s speed] [-c millisec] [-c nanosec]\n"
+           "[-r repeat] [-t ttl] [-d dest ipv4] [-g dest ipv6] [-p dest port] "
            "pcap\n"
            "\n"
            "  -i iface    interface to send packets through\n"
            "  -l          enable loopback\n"
            "  -c millisec constant milliseconds between packets\n"
+           "  -n nanosec constant nanoseconds between packets\n"
            "  -r repeat   number of times to loop data (-1 for infinite loop)\n"
            "  -s speed    replay speed relative to pcap timestamps\n"
            "  -t ttl      packet ttl\n"
@@ -211,15 +231,19 @@ int main(int argc, char *argv[]) {
       }
       auto udp = reinterpret_cast<const udphdr *>(p + sizeof(ether_header) +
                                                   ip->ip_hl * 4);
-      if (interval != -1) {
+      if(interval != -1){ // millisec precision
         // Use constant packet rate
         deadline.tv_sec += interval / 1000L;
         deadline.tv_nsec += (interval * 1000000L) % NANOSECONDS_PER_SECOND;
-      } else {
+      }else if(nano_interval != -1){ // nanosec precision
+        // Use constant packet rate
+        deadline.tv_sec += nano_interval / 1000000000L;
+        deadline.tv_nsec += (nano_interval * 1L) % NANOSECONDS_PER_SECOND;
+      }else{ // packet-based timing
         // Next packet deadline = start + (packet ts - first packet ts) * speed
         int64_t delta =
             (header.ts.tv_sec - pcap_start.tv_sec) * NANOSECONDS_PER_SECOND +
-            (header.ts.tv_usec -
+             (header.ts.tv_usec -
              pcap_start.tv_nsec); // Note PCAP_TSTAMP_PRECISION_NANO
         if (speed != 1.0) {
           delta *= speed;
@@ -235,12 +259,14 @@ int main(int argc, char *argv[]) {
         deadline.tv_nsec -= NANOSECONDS_PER_SECOND;
       }
 
-      timespec now = {};
+     // get the current time
+     timespec now = {};
       if (clock_gettime(CLOCK_MONOTONIC, &now) == -1) {
         std::cerr << "clock_gettime: " << strerror(errno) << std::endl;
         return 1;
       }
 
+      // works for either millisec or nanosec-based deadline
       if (deadline.tv_sec > now.tv_sec ||
           (deadline.tv_sec == now.tv_sec && deadline.tv_nsec > now.tv_nsec)) {
         if (clock_nanosleep(CLOCK_MONOTONIC, TIMER_ABSTIME, &deadline,
-- 
2.29.0

